File: mak/nokotan/norm.cpp
#include <iostream>
  
using namespace std;

const int SIZE = 4;

float normL1(float[]); // L1
float normL2(float[]); // L2 
float l_inf_norm(float[]); // L inf for max

int main() {
    float x[SIZE] = { 4, 4, 4, 4};
       
    cout << "Norm of the array (L1): " << normL1(x) << endl;
    cout << "Norm of the array (L2): " << normL2(x) << endl;
    cout << "Norm of the array (Linf): " << l_inf_norm(x) << endl;

    return EXIT_SUCCESS;
}

float normL1(float arr[]) {
    float sum = 0.0;
    for (int i = 0; i < SIZE; ++i)
        sum += abs(arr[i]); // Модуль каждого элемента

    return sum;
}

float normL2(float arr[]) {
    float sum = 0.0;
    for (int i = 0; i < SIZE; ++i) 
        sum += (arr[i] * arr[i]); 

    return sqrt(sum); 
}

float l_inf_norm(float arr[]) {
    float max_value = 0.0;
    for (int i = 0; i < SIZE; ++i) {
        float abs_value = abs(arr[i]);
        if (abs_value > max_value) max_value = abs_value;
    }

    return max_value;
}

File: mak/rewritten/norm.f90
program norms
    implicit none
    integer, parameter :: SIZE = 4
    real :: x(SIZE) = [12.7, 4.4, 2.3, 5.5]
    
    ! Вывод результатов с помощью write
    write(*,'(A, F8.1)') "Norm of the array (L1): ", normL1(x, SIZE)
    write(*,'(A, F8.1)') "Norm of the array (L2): ", normL2(x, SIZE)
    write(*,'(A, F6.1)') "Norm of the array (Linf): ", l_inf_norm(x, SIZE)

contains
    ! Функция для L1 нормы (сумма абсолютных значений)
    real function normL1(arr, n)
        integer, intent(in) :: n
        real, intent(in) :: arr(n)  ! Размер массива передается как аргумент
        real :: sum
        integer :: i
    
        sum = 0.0
        do i = 1, n
            sum = sum + abs(arr(i))
        end do
        normL1 = sum
    end function normL1

    ! Функция для L2 нормы 
    real function normL2(arr, n)
        integer, intent(in) :: n
        real, intent(in) :: arr(n)  ! Размер массива передается как аргумент
        real :: sum
        integer :: i
    
        sum = 0.0
        do i = 1, n
            sum = sum + sqrt(arr(i) * arr(i))  
        end do
        normL2 = sum
    end function normL2 

    ! Функция для L∞ нормы (максимальное абсолютное значение)
    real function l_inf_norm(arr, n)
        integer, intent(in) :: n
        real, intent(in) :: arr(n)  ! Размер массива передается как аргумент
        real :: max_value, abs_value
        integer :: i
    
        max_value = abs(arr(1))
        do i = 1, n
            abs_value = abs(arr(i))
            if (abs_value > max_value) max_value = abs_value
        end do
        l_inf_norm = max_value
    end function l_inf_norm

end program norms